### ISSUE TITLE
[BUG]: ASM output is omitting some ASM

### ISSUE THREAD

**Issue by hanickadot** [compiler-explorer/compiler-explorer#5244]
### Describe the bug

If I explicitly disable debug info with -g0 the ASM output is uncat than the original

### Steps to reproduce

in CTRE with DFA branch use this code:

```c++
#include <ctre.hpp>

bool is_hexdec(std::string_view input) {
    return ctre::fast_match<"(?:[0-9a-f][0-9a-f])+">(input);
}
```

and create two clones of clang one with `-std=c++23 -O3` and second with `-std=c++23 -O3 -g0`

output is different (cut short)

### Expected behavior

the outputs should be same

### Reproduction link

https://compiler-explorer.com/z/bf9edveoz

### Screenshots

<img width="2153" alt="Screenshot 2023-07-07 at 15 45 58" src="https://github.com/compiler-explorer/compiler-explorer/assets/6557263/be908eb3-1667-48f9-8457-fe0cf1ca5a0d">

### Operating System

Mac OS 13.4.1

### Browser version

Safari 16.5.1

**Comment by jeremy-rifkin** [comment]
I believe this is library code filtering. With -g0 our asm parser doesn’t know what is library code and what is not, disabling library filtering in the compiler without -g0 seems to produce the correct output.

**Comment by hanickadot** [comment]
Oh it's a header only library. And everything is a template. Nothing should be hidden.

**Comment by partouf** [comment]
I'd hate to say it, but this behaviour is actually correct. We consider everything except something linked to the user's code "library" code. Thus with the library filter you can turn this off and on.
It used to be wonky and incorrect, but we did some fixes to it.

**Comment by hanickadot** [comment]
This makes me sad. I can for myself make library code visible, but people checking out CTRE won't know.

**Comment by hanickadot** [comment]
Not sure if this is related to this bug or not ...

https://compiler-explorer.com/z/4eT694Y33

filtering library function filters out even code which is not associated to any library at all:

<img width="1759" alt="Screenshot 2023-07-19 at 23 20 54" src="https://github.com/compiler-explorer/compiler-explorer/assets/6557263/baf47b30-5d7c-4de1-8f90-6abb6f7c68be">
<img width="1759" alt="Screenshot 2023-07-19 at 23 21 00" src="https://github.com/compiler-explorer/compiler-explorer/assets/6557263/7c8d47a6-6cdd-4c5d-b7c6-d5f4fa232d55">

**Comment by hanickadot** [comment]
^ as you can see function::~function is hidden even if it should be there

**Comment by jeremy-rifkin** [comment]
Thanks for the report, that looks like a bug. Partouf will be the expert but it looks like an issue with what debug symbols gcc is emitting:

-O0
```x86asm
function<int>::~function() [base object destructor]:                    # @function<int>::~function() [base object destructor]
.Lfunc_begin13:
        .loc    2 31 0 is_stmt 1                # example.cpp:31:0
        .cfi_startproc
        push    rbp
        .cfi_def_cfa_offset 16
        .cfi_offset rbp, -16
        mov     rbp, rsp
        .cfi_def_cfa_register rbp
        sub     rsp, 16
        mov     qword ptr [rbp - 8], rdi
        mov     rdi, qword ptr [rbp - 8]
        .loc    2 32 16 prologue_end            # example.cpp:32:16
        call    std::__n4861::coroutine_handle<function<int>::promise_type>::destroy() const
        .loc    2 34 5 epilogue_begin           # example.cpp:34:5
        add     rsp, 16
        pop     rbp
        .cfi_def_cfa rsp, 8
        ret
```

-O3
```x86asm
function<int>::~function() [base object destructor]:                    # @function<int>::~function() [base object destructor]
.Lfunc_begin2:
        .cfi_startproc
        .loc    3 242 53 prologue_end is_stmt 1 # /opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/14.0.0/../../../../include/c++/14.0.0/coroutine:242:53
        mov     rdi, qword ptr [rdi]
.Ltmp11:
        .loc    3 242 30 is_stmt 0              # /opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/14.0.0/../../../../include/c++/14.0.0/coroutine:242:30
        jmp     qword ptr [rdi + 8]             # TAILCALL
```

I think the assembly parser treats it as library code since that's the only locations it sees.

We had a bug recently after llvm changed how debug symbols are emitted that caused some user code to be filtered out when wrapped in an std::function, for example. In this case it might be hard to address, since it looks like there's nothing at all to indicate that it came from user code. But again partouf will know more :)

**Comment by partouf** [comment]
Right, so this is because the `handle.destroy();` call is being inlined, and the compiler with various transforms concludes it to only contain code from stdlib.

This is -- we think -- a result of the changes that came with this https://reviews.llvm.org/D147506

There are pros and cons for this, but truthfully; we're really only visualizing the info that the compiler gives us.

Now in light of the changes to clang (and also gcc, really) - I can imagine people's opinions starting to vary on this. But the library code filter is about 5 years old now and has been beneficial in hiding unnecessary junk. There's also a very good reason that this filter is optional and not default, just because people might want to see what's going on in functions that are used.

I'm all ears really how to differentiate relevancy, but it needs to be something that we can actually code up with the data that we are given.

**Comment by hanickadot** [comment]
> There's also a very good reason that this filter is optional and not fixed, just because people might want to see what's going on in functions that are used.

IMHO it should be opt-in, it's always surprising for me going from "why this is not here" to "oh this is just library symbols filter"

**Comment by hanickadot** [comment]
I found another problem with the filtering which confused me today:

https://compiler-explorer.com/z/5fj71EPqf

vs

https://compiler-explorer.com/z/8q1xrE6ff (with unused filtering disabled)

<img width="1698" alt="Screenshot 2023-10-06 at 9 08 30" src="https://github.com/compiler-explorer/compiler-explorer/assets/6557263/795f781d-ad80-40ad-a3ea-6879e5081a86">

It confused me because I thought it made the first array 8 items long.

**Comment by partouf** [comment]
> I found another problem with the filtering which confused me today:
> 
> https://compiler-explorer.com/z/5fj71EPqf
> 
> vs
> 
> https://compiler-explorer.com/z/8q1xrE6ff (with unused filtering disabled)
> 
> <img alt="Screenshot 2023-10-06 at 9 08 30" width="1698" src="https://user-images.githubusercontent.com/6557263/273131381-795f781d-ad80-40ad-a3ea-6879e5081a86.png">
> It confused me because I thought it made the first array 8 items long.

This is unrelated to the original issue, and this could *maybe* be fixed.

**Comment by OfekShilon** [comment]
I tried tackling this, but my fix clashed with an older bug - #2788.
In #2788 we wish to avoid filtering away jump-table data like:
```
revc(char):
        sub     r3, r0, #65
        cmp     r3, #19
        ldrls   pc, [pc, r3, asl #2]
        b       .L7
.L4:
        .word   .L8     // <--
        .word   .L7     // <--
        .word   .L6     // <--
        .word   .L7     // <--
        .word   .L7     // <--
...
```
and here we wish to *avoid* filtering away unused data entries like -
```
a:
  .long 1
  .long 2
  .long 3
  .long 4
b:
  .long 1     // <--
  .long 2     // <--
  .long 3     // <--
  .long 4     // <--
```

To distinguish the cases we need to be perform much deeper analysis, and be able to deduce that `ldrls   pc, [pc, r3, asl #2]` is in fact a jump that might reach `.L4` - which seems unfeasible.

I think the least-confusing thing to do would be to always display labels whose contents consist entirely of data - as detected by https://github.com/compiler-explorer/compiler-explorer/blob/a08f2306881d436c90ea9bc30c6c4afe3d334863/lib/parsers/asm-parser.ts#L269
I.e., accept that for such labels we cannot detect whether they're used and display them always, regardless of filters.

Sounds ok to you?    @partouf   @mattgodbolt  @jeremy-rifkin

**Comment by partouf** [comment]
You could maybe use the `isUserFunctionByLookingAhead()` approach, but then for data. But I suspect you'll get very noisy assembly. Perhaps there should also be a way to optionally switch all data off?

**Comment by OfekShilon** [comment]
> You could maybe use the `isUserFunctionByLookingAhead()` approach, but then for data. But I suspect you'll get very noisy assembly (and you can't see if the data is yours or of some library). 

Generally I'd argue that hiding relevant lines is *more* confusing than displaying irrelevant lines (like library data). The #2788 example is a counter-data-point to this general rule, but we should be able to display `b` label with its contents it wouldn't be that confusing.
Also my uninformed opinion is that most 'noise' labels would contain some directives and not just data - and so would still be filtered out - but I might be very wrong.


> Perhaps there should also be a way to optionally filter out all data?

Sounds like a good idea, today I think data filtering is not controlled separately but in some messy way tangled to directives/labels.
But this can be handled separately from this issue.
